src/core/server/server.h

```diff
@@ -113,16 +146,88 @@ class Server : public ServerInterface,
+  /// Implementations must override the OrphanImpl() method, which should stop
   /// listening and initiate destruction of the listener.
   class ListenerInterface : public InternallyRefCounted<ListenerInterface> {
    public:
+    class ConfigFetcherWatcher
+        : public grpc_server_config_fetcher::WatcherInterface {
+     public:
+      explicit ConfigFetcherWatcher(RefCountedPtr<ListenerInterface> listener)
+          : listener_(std::move(listener)) {}
+
+      void UpdateConnectionManager(
+          RefCountedPtr<grpc_server_config_fetcher::ConnectionManager>
+              connection_manager) override;
+
+      void StopServing() override;
+
+     private:
+      const RefCountedPtr<ListenerInterface> listener_;
+    };
+
+    // State for a connection that is being managed by this listener.
+    class LogicalConnection : public InternallyRefCounted<LogicalConnection> {
+     public:
+      explicit LogicalConnection(RefCountedPtr<ListenerInterface> listener)
+          : listener_(std::move(listener)),
+            event_engine_(
+                listener_->server_->channel_args()
+                    .GetObject<
+                        grpc_event_engine::experimental::EventEngine>()) {}
+      LogicalConnection(const LogicalConnection&) = delete;
+      LogicalConnection& operator=(const LogicalConnection&) = delete;
+      LogicalConnection(LogicalConnection&&) = delete;
+      LogicalConnection& operator=(LogicalConnection&&) = delete;
+      ~LogicalConnection() override = default;
+
+      template <typename T>
+      RefCountedPtr<T> RefAsSubclass() {
+        return InternallyRefCounted<LogicalConnection>::RefAsSubclass<T>();
+      }
+
+      void SendGoAway();
+
+      // Cancel drain grace timer. It won't be started in the future either.
+      void CancelDrainGraceTimer();
+
+     protected:
+      grpc_event_engine::experimental::EventEngine* event_engine() const {
+        return event_engine_;
+      }
+
+      const RefCountedPtr<ListenerInterface>& listener() const {
+        return listener_;
+      }
+
+     private:
+      // The following two methods are called in the context of a server config
+      // event.
+
+      // Returns true if the operation was successful, false, if it was
+      // not performed for example if the connection was already shutdown.
+      virtual bool SendGoAwayImpl() = 0;
+      virtual void DisconnectImmediatelyImpl() = 0;
+
+      const RefCountedPtr<ListenerInterface> listener_;
+      grpc_event_engine::experimental::EventEngine* const event_engine_;
+      mutable Mutex mu_;
+      grpc_event_engine::experimental::EventEngine::TaskHandle
+          drain_grace_timer_handle_ ABSL_GUARDED_BY(&mu_) = grpc_event_engine::
+              experimental::EventEngine::TaskHandle::kInvalid;
+      // The drain grace timer should only be started if it wasn't previously
+      // cancelled.
+      bool drain_grace_timer_handle_cancelled_ ABSL_GUARDED_BY(&mu_) = false;
```

Comment:
Ah, since the base class no longer has the state that is now in the implementation, `drain_grace_timer_handle_cancelled_` is a way for the base class to know whether the transport already got closed.

(It is possible for us to have gotten the drain grace timer cancellation before we attempt to send the goaway.)